src/Data/Array/Destination.hs

```diff
@@ -2,9 +2,86 @@
+-- and DPS (destination passing style).
+--
+-- == What are destination arrays and DPS?
+--
+-- There are several lists which must, despite program optimizations
+-- (i.e., [deforesting](https://www.sciencedirect.com/science/article/pii/030439759090147A)),
+-- be allocated, filled, passed along and de-allocated. When the allocation
+-- of these arrays is controlled by the programmer and not
+-- done by Haskell's GC (garbage collector), programs are often more efficient.
+--
+-- DPS is a style of programming in which functions that produce such arrays
+-- take them as arguments -- moving the responsibility of allocating, using
+-- and freeing those arrays to the caller. The arrays passed as arguments
+-- to functions that must be written to are called /destination arrays/.
+--
+-- Here is a C function that adds vectors written in DPS style; it
+-- takes in the destiniation array @sum@ and writes to it:
+--
+-- @
+-- void add(int size, int *a, int *b, int *sum){
+--   for (int i=0; i<size;++i){sum[i]=a[i]+b[i];}
+-- }
+-- @
+--
+-- == Linear Types Enforce Destination Arrays are Filled
+--
+-- Linear types are used to ensure that the destination array
+-- is always written to. Why? Well:
+--
+-- Because of linear types, any function that uses a destination array
+-- must take that 'DArray' as a linear argument. This is because the only
+-- way to create a 'DArray' is through 'alloc' which takes a linear
+-- continutation of type @DArray a #-> ()@ and any function inside that
+-- continuation that uses a @DArray@ must use it linearly.
+--
+-- Since 'DArray' doesn't have a 'Consumable' instance, the only way to
+-- consume it is with the given API (e.g., with 'fill' or perhaps
+-- 'fromFunction') which fills the destiniation array completely
+-- (or errors as we desire).
+--
+-- For a deeper background on destination arrays see
+-- [this paper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/11/dps-fhpc17.pdf).
+--
+-- == Example
+--
+-- > import qualified Prelude.Linear as Linear
+-- > import qualified Data.Array.Destination as DPS
+-- > import qualified Data.Vector (Vector)
+-- > import qualified Data.Vector as Vector
+-- >
+-- > -- | Computes the diff of a vector and writes to an array
+-- > -- diff [1,2,3,4] is [2-1,3-2,4-3]
+-- > computeDiff :: Vector Int -> DPS.DArray Int #-> ()
+-- > computeDiff vec = DPS.fromFunction
+-- >   (\ix -> (vec Vector.! (ix + 1)) - (vec Vector.! ix))
+-- >
+-- > someAction :: IO ()
+-- > someAction = do
+-- >   vec <- inputVector
+-- >   let diffSize = (Vector.length vec) - 1
+-- >   let vecToSend = DPS.alloc diffSize (computeDiff vec)
+-- >   sendVectorToTwoServers vecToSend
+-- >   where
+-- >
+-- >     sendVectorToTwoServers :: Vector Int -> IO ()
+-- >     sendVectorToTwoServers _ = {- fake implementation -}
+-- >       return ()
+-- >
+-- >     inputVector :: IO (Vector Int)
+-- >     inputVector =
+-- >       return (fromList (map (\x -> (7 * (x+3)) `div` 11) [1..100]))
```

Comment:
Instead of describing an IO action, I think this would be a shorter example if we were to just describe a `Vector Int -> Vector Int` function. eg.

```
computeDiff :: Vector Int -> DPS.DArray Int #-> ()
computeDiff = undefined

vectorDiff :: Vector Int -> Vector Int
vectorDiff vec =
  let diffSize = (Vector.length vec) - 1
  in  DPS.alloc diffSize (computeDiff vec)
```